anti-diff/src/Data/FingerTree/RootMeasured/Strict.hs

```diff
@@ -105,147 +107,172 @@ infixl 5 |>
+        in  (vrLeft, invert vrLeft <> vr)
+
+-- | /O(f(l, r)) = O(length(r))/. Compute the root measures by subtraction of
+-- the right part's root measure.
+pattern SplitRight :: RootMeasured vr a => SplitRootMeasure vr vi a
+pattern SplitRight <- SplitRootMeasure _
+  where
+    SplitRight = SplitRootMeasure $
+      \vr (_left, right) ->
+        let vrRight = foldMap measureRoot right
+        in  (vr <> invert vrRight, vrRight)
+
+-- | /O(f(l, r))/. A custom function for computing root measures of a split.
+pattern SplitCustom ::
+     (  vr
+     -> ( FT.StrictFingerTree vi a
+        , FT.StrictFingerTree vi a
+        )
+     -> (vr, vr)
      )
-splitlr = \case
-  L -> splitl
-  R -> splitr
+  -> SplitRootMeasure vr vi a
+pattern SplitCustom f = SplitRootMeasure f
 
--- | Like @'split'@, but we compute root measures for /free/ through subtraction
--- of the left part's root measure.
---
--- This function is more performant than @'splitr'@ if the left part of the
--- split is shorter than the right part.
+instance RootMeasured vr a => Show (SplitRootMeasure vr vi a) where
+  show SplitLeft       = "SplitLeft"
+  show SplitRight      = "SplitRight"
+  show (SplitCustom _) = "SplitCustom"
+
+-- | /O(log(min(i,n-i))) + O(f(l, r)) = O(log(min(i,n-i))) + O(length(l))/.
+-- Specialisation of @'split'@ to use @'SplitLeft'@.
 --
--- TODO(jdral): Complexity analysis.
+-- Note: The @l@ suffix of the function name indicates that its time complexity
+-- depends on the length of the left part of the split.
 splitl ::
-     ( SuperMeasured vt vi a
-     , Group vt
-     )
+     SuperMeasured vr vi a
   => (vi -> Bool)
-  -> StrictFingerTree vt vi a
-  -> ( StrictFingerTree vt vi a
-     , StrictFingerTree vt vi a
+  -> StrictFingerTree vr vi a
+  -> ( StrictFingerTree vr vi a
+     , StrictFingerTree vr vi a
      )
-splitl p = split p f
-  where
-    f vt (vtLeft, _vtRight) = (vtLeft, invert vtLeft <> vt)
+splitl p = split p SplitLeft
 
--- | Like @'split'@, but we compute root measures for /free/ through subtraction
--- of the right part's root measure.
+-- | /O(log(min(i,n-i))) + O(f(l, r)) = O(log(min(i,n-i))) + O(length(r))/.
+-- Specialisation of @'split'@ to use @'SplitRight'@.
 --
--- This function is more performant than @'splitl'@ if the right part of the
--- split is shorter than the left part.
---
--- TODO(jdral): Complexity analysis.
+-- Note: The @r@ suffix of the function name indicates that its time complexity
+-- depends on the length of the right part of the split.
 splitr ::
-     ( SuperMeasured vt vi a
-     , Group vt
-     )
+     SuperMeasured vr vi a
   => (vi -> Bool)
-  -> StrictFingerTree vt vi a
-  -> ( StrictFingerTree vt vi a
-     , StrictFingerTree vt vi a
+  -> StrictFingerTree vr vi a
+  -> ( StrictFingerTree vr vi a
+     , StrictFingerTree vr vi a
      )
-splitr p = split p f
-  where
-    f vt (_vtLeft, vtRight) = (vt <> invert vtRight, vtRight)
-
+splitr p = split p SplitRight
```

Comment:
@Jasagredo @dnadales what do you think of this re-organisation of the split functions? The documentation should now also comment on the time complexity of these functions.

These changes were made in reponse to:
* https://github.com/input-output-hk/ouroboros-network/pull/3997#discussion_r972911366
* https://github.com/input-output-hk/ouroboros-network/pull/3997#discussion_r972909487